### PR TITLE
PoA_DynamicFederationMembers_4

### PR DESCRIPTION
**Pull request by noescape00**
VotingManaer, PollResultExecutor, api controller

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.PoA/Voting/PollResultExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+﻿using NBitcoin;
+
+namespace Stratis.Bitcoin.Features.PoA.Voting
+{
+    public interface IPollResultExecutor
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.PoA/Voting/PollResultExecutor.cs]: The policy requires at least a high level description of this interface and methods

File: src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs [hunk 2]
```diff
@@ -209,14 +222,74 @@ private void onBlockConnected(ChainedHeaderBlock chBlock)
         }
 
         // TODO subscribe to signals
-        private void onBlockDisconnected(ChainedHeaderBlock chBlock)
+        // TODO make private when signaler PR is merged
+        public void onBlockDisconnected(ChainedHeaderBlock chBlock)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs]: Such a big benefit in using `onBlockDisconnected` no need to deal with reorgs

File: src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs [hunk 3]
```diff
@@ -209,14 +222,74 @@ private void onBlockConnected(ChainedHeaderBlock chBlock)
...
+                this.logger.LogTrace("(-)[NO_VOTING_DATA]");
+                return;
+            }
+
+            lock (this.locker)
+            {
+                foreach (VotingData votingData in votingDataList)
+                {
+                    // Poll that was finished in the block being disconnected.
+                    Poll pollToRevert = this.finishedPolls.SingleOrDefault(x => x.PollAppliedBlockHash == chBlock.Block.GetHash() && x.VotingData == votingData);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs]: `chBlock.Block.GetHash() ` is the `ChainedHeader.HashBlock`

File: src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs [hunk 4]
```diff
@@ -209,14 +222,74 @@ private void onBlockConnected(ChainedHeaderBlock chBlock)
...
+        {
+            int largestId = -1;
+
+            foreach (Poll finishedPoll in this.finishedPolls)
+            {
+                if (finishedPoll.Id > largestId)
+                    largestId = finishedPoll.Id;
+            }
+
+            foreach (Poll pendingPoll in this.pendingPolls)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs]: This means after a reorg a pool id will be reused right? which I think is correct

File: src/Stratis.Bitcoin.Features.PoA/Voting/PollResultExecutor.cs (hunk 1 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA/Voting/PollResultExecutor.cs]: Ok, will comment it in the next PR

File: src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs (hunk 3 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs]: yeah
and both of them will returned cached value

File: src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs (hunk 4 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs]: yes. order will be restored and we need IDs for it.

**Pull request merged**